### ISSUE TITLE
An error when nomad self-bootstrap cluster with consul.Thanks

### ISSUE THREAD

**Issue by cgyim1992** [hashicorp/nomad#1792]
### Nomad version

 `nomad version 0.4.1`
### Operating system and Environment details

```
Ubuntu14.04,Consul version 0.7.0  
```
### Issue

I am new to this . I have 3 VMs,want to set up a nomad cluster(all 3 are servers) with consul cluster and start with supervisor,but I get some errors ,the consul cluster seems successfully set up ,but nomad cluster failed.the commmand `consul members` output the correct cluster member and all alive ,but command
`nomad server-members`output only one ,the server itself,the cluster failed to self-bootstrap. 
**What's more, I can manually join nomad cluster with `nomad server-join  x.x.x.x`.And after join these 3 servers in a cluster, nomad election works, it has a leader.The problem is the nomad cluster can not self-boostrap with consul.**It seems nomad server can not detect the other servers but consul cluster was set up.
VM ip are `172.16.30.143`,`172.16.30.144`,`172.16.30.145`.All server mode.
### `172.16.30.143`

agent:
    check_monitors = 0
    check_ttls = 0
    checks = 4
    services = 4
build:
    prerelease =
    revision = 'a189091
    version = 0.7.0
consul:
    bootstrap = false
    known_datacenters = 1
    leader = false
    leader_addr = 172.16.30.145:8300
    server = true
raft:
    applied_index = 4310
    commit_index = 4310
    fsm_pending = 0
    last_contact = 47.660331ms
    last_log_index = 4310
    last_log_term = 8
    last_snapshot_index = 0
    last_snapshot_term = 0
    latest_configuration = [{Suffrage:Voter ID:172.16.30.143:8300 Address:172.16.30.143:8300} {Suffrage:Voter ID:172.16.30.145:8300 Address:172.16.30.145:8300} {Suffrage:Voter ID:172.16.30.144:8300 Address:172.16.30.144:8300}]
    latest_configuration_index = 82
    num_peers = 2
    protocol_version = 1
    protocol_version_max = 3
    protocol_version_min = 0
    snapshot_version_max = 1
    snapshot_version_min = 0
    state = Follower
    term = 8
### `172.16.30.144`

agent:
    check_monitors = 0
    check_ttls = 0
    checks = 4
    services = 4
build:
    prerelease =
    revision = 'a189091
    version = 0.7.0
consul:
    bootstrap = false
    known_datacenters = 1
    leader = false
    leader_addr = 172.16.30.145:8300
    server = true
raft:
    applied_index = 4321
    commit_index = 4321
    fsm_pending = 0
    last_contact = 49.821475ms
    last_log_index = 4321
    last_log_term = 8
    last_snapshot_index = 0
    last_snapshot_term = 0
    latest_configuration = [{Suffrage:Voter ID:172.16.30.143:8300 Address:172.16.30.143:8300} {Suffrage:Voter ID:172.16.30.145:8300 Address:172.16.30.145:8300} {Suffrage:Voter ID:172.16.30.144:8300 Address:172.16.30.144:8300}]
    latest_configuration_index = 82
    num_peers = 2
    protocol_version = 1
    protocol_version_max = 3
    protocol_version_min = 0
    snapshot_version_max = 1
    snapshot_version_min = 0
    state = Follower
    term = 8         
### `172.16.30.145`

 agent:
    check_monitors = 0
    check_ttls = 0
    checks = 4
    services = 4
build:
    prerelease =
    revision = 'a189091
    version = 0.7.0
consul:
    bootstrap = false
    known_datacenters = 1
    leader = true
    leader_addr = 172.16.30.145:8300
    server = true
raft:
    applied_index = 4325
    commit_index = 4325
    fsm_pending = 0
    last_contact = never
    last_log_index = 4325
    last_log_term = 8
    last_snapshot_index = 0
    last_snapshot_term = 0
    latest_configuration = [{Suffrage:Voter ID:172.16.30.143:8300 Address:172.16.30.143:8300} {Suffrage:Voter ID:172.16.30.145:8300 Address:172.16.30.145:8300} {Suffrage:Voter ID:172.16.30.144:8300 Address:172.16.30.144:8300}]
    latest_configuration_index = 82
    num_peers = 2
    protocol_version = 1
    protocol_version_max = 3
    protocol_version_min = 0
    snapshot_version_max = 1
    snapshot_version_min = 0
    state = Leader
    term = 8  
### Supervisor consul config

[program:consul]
command=consul agent -config-dir /etc/consul.d
directory=/tmp/
stdout_logfile=/data/logs/supervisor/consul.out
stderr_logfile=/data/logs/supervisor/consul.err
user=root
autorestart=true
autostart=true
startsecs=5
startretries=3
stopwaitsecs=5
### Supervisor nomad config

[program:nomad]
command=nomad agent -config /etc/nomad/server.hcl
directory=/tmp/
stdout_logfile=/data/logs/supervisor/nomad.out
stderr_logfile=/data/logs/supervisor/nomad.err
user=root
autostart=true
autorestart=true
startsecs=5
startretries=3
stopwaitsecs=5
### Consul server config(config.json)  ip:`172.16.30.144`

{"datacenter": "dc1",
  "data_dir": "/data/consul",
  "log_level": "INFO",
  "node_name": "nomad02",
  "server": true,
  "advertise_addr": "172.16.30.144",
  "addresses": {
    "http": "0.0.0.0"
  },
  "ui":true,
  "ports": {
    "https": -1
  },
  "check": {
    "id": "dc1-check",
    "name": "http on port 8500",
    "http": "http://172.16.30.144:8500/",
    "interval": "5s",
    "timeout": "1s"
  },
  "start_join": ["172.16.30.144"],
  "retry_join":["172.16.30.143","172.16.30.145"],
  "bootstrap_expect":3,
  "retry_interval": "30s"
}

The other 2 VM are similar except ip address and join address and node name
### Nomad Server config(server.hcl) ip:`172.16.30.144`

region = "yzjcluster"
datacenter = "dc1"
name = "nomad02"
data_dir = "/data/nomad/"
log_level = "DEBUG"
          # enable_syslog = true
          # syslog_facility = "LOCAL7"
bind_addr = "0.0.0.0"
          #bind_addr = "10.0.0.1"
advertise{
    http = "172.16.30.144:4646"
    rpc = "172.16.30.144:4647"
    serf = "172.16.30.144:4648"
}
leave_on_interrupt = true
leave_on_terminate = true
enable_debug = true
disable_update_check = true
consul {
    address = "127.0.0.1:8500"
    server_service_name = "yzjserver_service02"
    client_service_name = "yzjclient_service02"
    auto_advertise = true
    server_auto_join = true
    client_auto_join = true  
    ssl = false
    verify_ssl = false
}
           # Ship metrics to monitor the health of the cluster and to see task resource
           # usage.
           #telemetry {
           #  statsite_address = "172.16.30.100:8125"
           #    disable_hostname = false
           #    publish_allocation_metrics = true
           #    publish_node_metrics = true
           #}

server {
     enabled = true
     bootstrap_expect = 3
    num_schedulers = 2
    node_gc_threshold = "10m"
    rejoin_after_leave = true
     retry_max = 10
}
### The IP address is `172.16.30.145`

```
2016/10/05 16:30:48 [INFO] raft: Node at 172.16.30.145:4647 [Follower] entering Follower state (Leader: "")
    2016/10/05 16:30:48 [INFO] serf: EventMemberJoin: nomad03.yzjcluster 172.16.30.145
    2016/10/05 16:30:48.506081 [INFO] nomad: starting 2 scheduling worker(s) for [service batch system _core]
    2016/10/05 16:30:48.508028 [INFO] nomad: adding server nomad03.yzjcluster (Addr: 172.16.30.145:4647) (DC: dc1)
    2016/10/05 16:30:49 [WARN] raft: EnableSingleNode disabled, and no known peers. Aborting election.
    2016/10/05 16:30:54.245946 [DEBUG] server.consul: lost contact with Nomad quorum, falling back to Consul for server list
    2016/10/05 16:30:54.247727 [DEBUG] consul.syncer: error in syncing: 1 error(s) occurred:

* 1 error(s) occurred:

* Put http://127.0.0.1:8500/v1/agent/service/register: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
    2016/10/05 16:30:54.249267 [DEBUG] server.consul: lost contact with Nomad quorum, falling back to Consul for server list
    2016/10/05 16:30:54.250710 [DEBUG] consul.syncer: syncs succesful
    2016/10/05 16:30:55 [DEBUG] memberlist: TCP connection from=172.16.30.145:18496
    2016/10/05 16:30:59.204623 [ERR] worker: failed to dequeue evaluation: No cluster leader
    2016/10/05 16:30:59.307544 [ERR] worker: failed to dequeue evaluation: No cluster leader
    2016/10/05 16:31:00.648615 [ERR] http: Request /v1/status/peers, error: No cluster leader
    2016/10/05 16:31:00.649698 [DEBUG] http: Request /v1/status/peers (5.109543323s)
    2016/10/05 16:31:05 [DEBUG] memberlist: TCP connection from=172.16.30.145:18504
    2016/10/05 16:31:06.244761 [ERR] worker: failed to dequeue evaluation: No cluster leader
    2016/10/05 16:31:06.399137 [ERR] worker: failed to dequeue evaluation: No cluster leader
    2016/10/05 16:31:13.599718 [ERR] http: Request /v1/status/peers, error: No cluster leader
    2016/10/05 16:31:13.600302 [DEBUG] http: Request /v1/status/peers (5.059785325s)
    2016/10/05 16:31:15 [DEBUG] memberlist: TCP connection from=172.16.30.145:18516
    2016/10/05 16:31:19.402703 [ERR] worker: failed to dequeue evaluation: No cluster leader
    2016/10/05 16:31:19.500528 [ERR] worker: failed to dequeue evaluation: No cluster leader
    2016/10/05 16:31:25 [DEBUG] memberlist: TCP connection from=172.16.30.145:18526
    2016/10/05 16:31:26.593106 [ERR] http: Request /v1/status/peers, error: No cluster leader
    2016/10/05 16:31:26.593330 [DEBUG] http: Request /v1/status/peers (5.051045653s)
    2016/10/05 16:31:34.521322 [ERR] worker: failed to dequeue evaluation: No cluster leader
    2016/10/05 16:31:34.528811 [ERR] worker: failed to dequeue evaluation: No cluster leader
    2016/10/05 16:31:35 [DEBUG] memberlist: TCP connection from=172.16.30.145:18536
    2016/10/05 16:31:39.694254 [ERR] http: Request /v1/status/peers, error: No cluster leader
    2016/10/05 16:31:39.694840 [DEBUG] http: Request /v1/status/peers (5.151958335s)
    2016/10/05 16:31:45 [DEBUG] memberlist: TCP connection from=172.16.30.145:18546
    2016/10/05 16:31:49.576327 [ERR] worker: failed to dequeue evaluation: No cluster leader
    2016/10/05 16:31:49.754929 [ERR] worker: failed to dequeue evaluation: No cluster leader
    2016/10/05 16:31:50.875099 [DEBUG] server.consul: lost contact with Nomad quorum, falling back to Consul for server list
    2016/10/05 16:31:52.555437 [ERR] http: Request /v1/status/peers, error: No cluster leader
    2016/10/05 16:31:52.555665 [DEBUG] http: Request /v1/status/peers (5.011558017s)
    2016/10/05 16:31:55 [DEBUG] memberlist: TCP connection from=172.16.30.145:18556
    2016/10/05 16:31:59.624949 [DEBUG] http: Request /v1/agent/members (25.463419ms)
### Nomad Client logs (if appropriate)
### The IPaddress is `172.16.30.145`
    2016/10/05 17:27:18 [INFO] raft: Initial configuration (index=82): [{Suffrage:Voter ID:172.16.30.143:8300 Address:172.16.30.143:8300} {Suffrage:Voter ID:172.16.30.145:8300 Address:172.16.30.145:8300} {Suffrage:Voter ID:172.16.30.144:8300 Address:172.16.30.144:8300}]
    2016/10/05 17:27:18 [INFO] raft: Node at 172.16.30.145:8300 [Follower] entering Follower state (Leader: "")
    2016/10/05 17:27:18 [INFO] serf: EventMemberJoin: nomad03 172.16.30.145
    2016/10/05 17:27:18 [INFO] serf: EventMemberJoin: nomad03.dc1 172.16.30.145
    2016/10/05 17:27:18 [INFO] serf: Attempting re-join to previously known node: nomad02: 172.16.30.144:8301
    2016/10/05 17:27:18 [INFO] consul: Adding LAN server nomad03 (Addr: tcp/172.16.30.145:8300) (DC: dc1)
    2016/10/05 17:27:18 [INFO] consul: Raft data found, disabling bootstrap mode
    2016/10/05 17:27:18 [INFO] consul: Adding WAN server nomad03.dc1 (Addr: tcp/172.16.30.145:8300) (DC: dc1)
    2016/10/05 17:27:18 [INFO] serf: Attempting re-join to previously known node: server03@145.dc1: 172.16.30.145:8302
    2016/10/05 17:27:18 [WARN] serf: Failed to re-join any previously known node
    2016/10/05 17:27:18 [INFO] serf: Re-joined to previously known node: server03@145.dc1: 172.16.30.145:8302
    2016/10/05 17:27:18 [WARN] Service name "yzjserver_service03" will not be discoverable via DNS due to invalid characters. Valid characters include all alpha-numerics and dashes.
    2016/10/05 17:27:18 [WARN] Service name "yzjserver_service03" will not be discoverable via DNS due to invalid characters. Valid characters include all alpha-numerics and dashes.
    2016/10/05 17:27:18 [WARN] Service name "yzjserver_service03" will not be discoverable via DNS due to invalid characters. Valid characters include all alpha-numerics and dashes.
    2016/10/05 17:27:18 [INFO] agent: (LAN) joining: [172.16.30.145]
    2016/10/05 17:27:18 [INFO] agent: (LAN) joined: 1 Err: <nil>
    2016/10/05 17:27:18 [INFO] agent: Joining cluster...
    2016/10/05 17:27:18 [INFO] agent: (LAN) joining: [172.16.30.143 172.16.30.144]
    2016/10/05 17:27:18 [INFO] agent: (LAN) joined: 0 Err: 2 error(s) occurred:

* Failed to join 172.16.30.143: dial tcp 172.16.30.143:8301: getsockopt: connection refused
* Failed to join 172.16.30.144: dial tcp 172.16.30.144:8301: getsockopt: connection refused
    2016/10/05 17:27:18 [WARN] agent: Join failed: 2 error(s) occurred:

* Failed to join 172.16.30.143: dial tcp 172.16.30.143:8301: getsockopt: connection refused
* Failed to join 172.16.30.144: dial tcp 172.16.30.144:8301: getsockopt: connection refused, retrying in 30s
    2016/10/05 17:27:23 [INFO] serf: EventMemberJoin: nomad02 172.16.30.144
    2016/10/05 17:27:23 [INFO] consul: Adding LAN server nomad02 (Addr: tcp/172.16.30.144:8300) (DC: dc1)
    2016/10/05 17:27:25 [ERR] agent: failed to sync remote state: No cluster leader
    2016/10/05 17:27:27 [WARN] raft: Heartbeat timeout from "" reached, starting election
    2016/10/05 17:27:27 [INFO] raft: Node at 172.16.30.145:8300 [Candidate] entering Candidate state in term 8
    2016/10/05 17:27:27 [ERR] raft: Failed to make RequestVote RPC to {Voter 172.16.30.143:8300 172.16.30.143:8300}: dial tcp 172.16.30.143:8300: getsockopt: connection refused
    2016/10/05 17:27:27 [WARN] agent: http request failed 'http://172.16.30.145:4646/v1/status/peers': Get http://172.16.30.145:4646/v1/status/peers: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
    2016/10/05 17:27:27 [INFO] raft: Election won. Tally: 2
    2016/10/05 17:27:27 [INFO] raft: Node at 172.16.30.145:8300 [Leader] entering Leader state
    2016/10/05 17:27:27 [INFO] raft: Added peer 172.16.30.143:8300, starting replication
    2016/10/05 17:27:27 [INFO] raft: Added peer 172.16.30.144:8300, starting replication
    2016/10/05 17:27:27 [INFO] consul: cluster leadership acquired
    2016/10/05 17:27:27 [INFO] consul: New leader elected: nomad03
    2016/10/05 17:27:27 [INFO] raft: pipelining replication to peer {Voter 172.16.30.144:8300 172.16.30.144:8300}
   2016/10/05 17:27:27 [INFO] serf: EventMemberJoin: nomad01 172.16.30.143
    2016/10/05 17:27:27 [INFO] consul: Adding LAN server nomad01 (Addr: tcp/172.16.30.143:8300) (DC: dc1)
    2016/10/05 17:27:27 [INFO] consul: member 'nomad01' joined, marking health alive
    2016/10/05 17:27:29 [INFO] agent: Synced check 'dc1-check'
    2016/10/05 17:27:40 [WARN] agent: http request failed 'http://172.16.30.145:4646/v1/status/peers': Get http://172.16.30.145:4646/v1/status/peers: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
    2016/10/05 17:27:48 [INFO] agent: (LAN) joining: [172.16.30.143 172.16.30.144]
    2016/10/05 17:27:48 [INFO] agent: (LAN) joined: 2 Err: <nil>
    2016/10/05 17:27:48 [INFO] agent: Join completed. Synced with 2 initial agents
    2016/10/05 17:27:53 [WARN] agent: http request failed 'http://172.16.30.145:4646/v1/status/peers': Get http://172.16.30.145:4646/v1/status/peers: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
    2016/10/05 17:28:06 [WARN] agent: http request failed 'http://172.16.30.145:4646/v1/status/peers': Get http://172.16.30.145:4646/v1/status/peers: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
    2016/10/05 17:28:19 [WARN] agent: http request failed 'http://172.16.30.145:4646/v1/status/peers': Get http://172.16.30.145:4646/v1/status/peers: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
    2016/10/05 17:28:32 [WARN] agent: http request failed 'http://172.16.30.145:4646/v1/status/peers': Get http://172.16.30.145:4646/v1/status/peers: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
    2016/10/05 17:28:45 [WARN] agent: http request failed 'http://172.16.30.145:4646/v1/status/peers': Get http://172.16.30.145:4646/v1/status/peers: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
    2016/10/05 17:28:58 [WARN] agent: http request failed 'http://172.16.30.145:4646/v1/status/peers': Get http://172.16.30.145:4646/v1/status/peers: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
    2016/10/05 17:29:11 [WARN] agent: http request failed 'http://172.16.30.145:4646/v1/status/peers': Get http://172.16.30.145:4646/v1/status/peers: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
    2016/10/05 17:29:24 [WARN] agent: http request failed 'http://172.16.30.145:4646/v1/status/peers': Get http://172.16.30.145:4646/v1/status/peers: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
    2016/10/05 17:29:37 [WARN] agent: http request failed 'http://172.16.30.145:4646/v1/status/peers': Get http://172.16.30.145:4646/v1/status/peers: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
    2016/10/05 17:29:50 [WARN] agent: http request failed 'http://172.16.30.145:4646/v1/status/peers': Get http://172.16.30.145:4646/v1/status/peers: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
    2016/10/05 17:30:03 [WARN] agent: http request failed 'http://172.16.30.145:4646/v1/status/peers': Get http://172.16.30.145:4646/v1/status/peers: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
    2016/10/05 17:30:16 [WARN] agent: http request failed 'http://172.16.30.145:4646/v1/status/peers': Get http://172.16.30.145:4646/v1/status/peers: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
    2016/10/05 17:30:29 [WARN] agent: http request failed 'http://172.16.30.145:4646/v1/status/peers': Get http://172.16.30.145:4646/v1/status/peers: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
    2016/10/05 17:30:42 [WARN] agent: http request failed 'http://172.16.30.145:4646/v1/status/peers': Get http://172.16.30.145:4646/v1/status/peers: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
    2016/10/05 17:30:55 [WARN] agent: http request failed 'http://172.16.30.145:4646/v1/status/peers': Get http://172.16.30.145:4646/v1/status/peers: net/http: request canceled (Client.Timeout exceeded while awaiting headers)

```

**Comment by swapnildahiphale** [comment]
facing same issue. Please update if you solved this

**Comment by schmichael** [comment]
Sorry for the slow reply, but I'm unable to reproduce this. Seeing the timeouts and connectivity errors in the logs makes me think there's a network misconfiguration. Are you sure 172.16.30.143, 172.16.30.144, and 172.16.30.144 are able to fully connect to each other on the following ports:

* 8500 (consul http)
* 8400 (consul rpc)
* 8301, 8302 (consul serf lan, wan)
* 8300 (consul server)
* 4646 (nomad http)
* 4647 (nomad rpc)
* 4648 (nomad serf)

I've tried to create a simple local cluster that mirrors the relevant parts of your configuration without being able to reproduce the issue.

I pushed my configs here: https://github.com/schmichael/1792-test
